core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java

```diff
@@ -175,29 +182,40 @@ protected Location onReplicaShard(IndexRequest request, IndexShard indexShard) {
     }
 
     public static WriteResult<IndexResponse> executeIndexRequestOnPrimary(IndexRequest request, IndexShard indexShard,
-            MappingUpdatedAction mappingUpdatedAction) throws Exception {
-        Engine.Index operation = prepareIndexOperationOnPrimary(request, indexShard);
+            MappingUpdatedAction mappingUpdatedAction) {
+        Engine.Index operation;
+        try {
+            operation = prepareIndexOperationOnPrimary(request, indexShard);
+        } catch (Exception e) {
+            return new WriteResult<>(e);
+        }
         Mapping update = operation.parsedDoc().dynamicMappingsUpdate();
         final ShardId shardId = indexShard.shardId();
         if (update != null) {
-            mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
-            operation = prepareIndexOperationOnPrimary(request, indexShard);
+            try {
+                mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
```

Comment:
Yes, due to timing out for example. Now we bubble the exception up to be handled in the replication action instead of capturing it